### ISSUE TITLE
Better handling of [TAB] (ASCII 9) character.

### ISSUE THREAD

**Issue by loren-osborn** [mintty/mintty#1182]
### `xterm` Compatibility Issue
(Specifically, there is an issue with **preserving** `xterm` behavior)

When a HT character is printed beyond the terminal's 79th character column the cursor is advanced to the end of the current line; resulting on the next text output to be line-broken at the second character.  This behavior **at least** violates the "Principle of least astonishment" if nothing else.

You could argue that you're emulating physical terminals, which were never wider than 80 columns, but that sounds like a cop-out.

I've personally found this behavior most annoyingly noticeable in the output of the Go programming language's built-in test suite, but I imagine outputting HT characters that far right in output is somewhat uncommon.

**Comment by mintty** [comment]
Duplicate of #1020 and see explanations there. It's certainly not about referring to 80 columns, behaviour is the same with any terminal width. It's about compatibilty. Should it be changed, it might break other software, so the fact that it is astonishing behaviour indeed does not necessarily imply it should be "fixed". Please also consider that there does not seem to be any terminal around that would handle this differently, for the reasons mentioned. If the Go system has a problem with that it's a Go bug, please report it there as they may be ignoring established terminal behaviour or making invalid assumptions about it.

**Comment by mintty** [comment]
And for these reasons it'c certainly not a bug. Please remove the introductory word.

**Comment by loren-osborn** [comment]
I completely agree that compatibility trumps all... I mentioned `xterm` solely because you did, but we agree it needs to behave the same as mainstream terminals.

"Distinct from" can mean different things. In this case I was referring to the **effects** of

 - having gaping "line-wrap like" spaces on the right side of the window, before a single character in the last column
 
being distinct from

 - having tabs in the last column being zero-width,

not their cause, which likely **is** shared.

The next time I'm in a non-proprietary go project with tests, I'll post a dump of the test output for your review. but I don't know if this will happen this week.  I will point out that `less` displays the content in the manner I would expect

**Because** I value backwards compatibility, that is why I suggested a configuration option to treat the first 80 columns specially (as is the current, compatible, behavior)... The non-compatible behavior would be to not treat the left 80 columns as the only screen-region with tab stops.

I understand, at this time, that we're simply just discussing options.

**Comment by j4james** [comment]
> Windows PowerShell displays the content in the manner I would expect

I'm not sure what exactly you are complaining about in this issue, but I do know there was a bug in the tab handling in the Windows console, which is quite likely still present if you're using an older version of Windows 10 (like pre Windows 11). So just be aware that what you're seeing there is probably not the way the Windows console works anymore, and may not be a great example to use as something that matches your expected behavior.

**Comment by mintty** [comment]
There is an issue with TAB when output in the "pending wrap" state, i.e. *after* the last column - the line will wrap on the next character but the TAB will be skipped. Output your example file at terminal width 89 to see it.
Again, same behaviour in xterm, but xterm has an optional workaround for it, see xterm manual page option `-cu` for more information. If that is your issue, I will consider a similar option.

**Comment by loren-osborn** [comment]
I'm still investigating this intermittently, but it appears to be an interaction between my bash shell prompt and window resizing the window... I'll include my prompt string here for reference, but I haven't exactly identified the issue exactly yet.

```
export $PS1='\[\e]0;\u@\h: \w\a\]${debian_chroot:+($debian_chroot)}\[\033[01;32m\]\u@\h\[\033[00m\]:\[\033[01;34m\]\w\[\033[00m\]\$ '
```

I will post again when I know more.

**Comment by mintty** [comment]
Released 3.6.3.
